### ISSUE TITLE
False positive unsubscriptable-object with MutableMapping / pygtrie

### ISSUE THREAD

**Issue by originalrkk** [pylint-dev/pylint#9179]
### Bug description

```python
test.py:

from collections.abc import MutableMapping
from typing import cast
import pygtrie

T = MutableMapping[str, dict[str, int]]

if __name__ == '__main__':
    x: T = pygtrie.CharTrie({'a': {'b': 'c'}})
    y: pygtrie.CharTrie[str, dict[str, int]] = pygtrie.CharTrie({'a': {'b': 'c'}})
    z: dict[str, dict[str, int]] = {'a': {'b': 'c'}}

    # x and y print() lines trigger unsubscriptable-object, but z does not:
    print(x['a']['b'])
    print(cast(dict[str, dict[str, int]], x['a'])['b'])
    print(cast(pygtrie.CharTrie[str, dict[str, int]], x['a'])['b'])
    print(y['a']['b'])
    print(cast(T, y['a'])['b'])
    print(cast(dict[str, dict[str, int]], y['a'])['b'])
    print(z['a']['b'])
    print(cast(T, z['a'])['b'])
    print(cast(pygtrie.CharTrie[str, dict[str, int]], z['a'])['b'])
```
```


### Configuration

_No response_

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:13:10: E1136: Value 'x['a']' is unsubscriptable (unsubscriptable-object)
test.py:14:10: E1136: Value 'cast(dict[str, dict[str, int]], x['a'])' is unsubscriptable (unsubscriptable-object)
test.py:15:10: E1136: Value 'cast(pygtrie.CharTrie[str, dict[str, int]], x['a'])' is unsubscriptable (unsubscriptable-object)
test.py:16:10: E1136: Value 'y['a']' is unsubscriptable (unsubscriptable-object)
test.py:17:10: E1136: Value 'cast(T, y['a'])' is unsubscriptable (unsubscriptable-object)
test.py:18:10: E1136: Value 'cast(dict[str, dict[str, int]], y['a'])' is unsubscriptable (unsubscriptable-object)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

With `MutableMapping[str, dict[str, int]]`, pylint should recognize that `x['a']` or `y['a']` will yield a dict, which is subscriptable.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.1
Python 3.11.3
```
